### PR TITLE
perf(router-core): streamline route initialization

### PR DESCRIPTION
**Pull request by schiller-manuel**
## 🎯 Changes

Reuse parsed route segments, simplify route-owned interpolation, stabilize path decoders, and initialize route trees directly with explicit indexes.

This is PR 12 of a stacked series and is based on `optimize-link-vue-match`.

## ✅ Checklist

- [x] I have followed the steps in the [Contributing guide](https://github.com/TanStack/router/blob/main/CONTRIBUTING.md).
- [x] I have tested code changes locally with the relevant test commands.
- [x] I fully understand the code in this pull request, including any code generated with AI assistance.

## 🚀 Release Impact

- [x] This change affects published code, and I have generated changesets.
- [ ] This change is docs/CI/dev-only.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved path parameter interpolation and encoding, including wildcard and optional parameters.
  * Preserved original parameter names and fallback route behavior across route-tree rebuilds.
  * Devtools now hides navigation actions when required path parameters are missing.

* **Performance**
  * Improved route-tree construction and path generation by reusing parsed route information.

* **Documentation**
  * Clarified that allowed path parameter characters must be configured when creating a router and cannot be changed afterward.

### CODE REVIEW

File: packages/router-core/src/new-process-route-tree.ts [hunk 1]
```diff
@@ -647,11 +601,6 @@ export function findRouteMatch<
   return result
 }
 
-/** Trim trailing slashes (except preserving root '/'). */
-export function trimPathRight(path: string) {
```
Comment (schiller-manuel) [inline packages/router-core/src/new-process-route-tree.ts:651]: uses the imported one instead that already existed

File: packages/router-core/tests/original-param-names.test.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+      { id: '/first', path: '$id/detail', fullPath: '/$id/detail' },
+      {
+        id: '/alias',
+        path: '$name',
+        fullPath: '/$name',
+        children: [
+          {
+            id: '/alias/detail',
+            path: 'detail',
+            fullPath: '/$name/detail',
```
Comment (Sheraff) [inline packages/router-core/tests/original-param-names.test.ts:41]: why are these route ids not related to their paths? (e.g. id `/first` and path `/$id`)
Is that supposed to be valid?

File: packages/router-core/tests/path-decoder.bench.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import { bench, describe, expect } from 'vitest'
+import { createMemoryHistory } from '@tanstack/history'
+import { BaseRootRoute, BaseRoute } from '../src'
+import { createTestRouter } from './routerTestUtils'
+
+if (process.env.TSR_LINK_PERF === '1') {
```
Comment (Sheraff) [inline packages/router-core/tests/path-decoder.bench.ts:6]: why is this behind a flag? there is not risk of executing it w/ the tests since it's a `.bench.ts` file, right?

Comment (schiller-manuel) [reply packages/router-core/tests/path-decoder.bench.ts:6]: it allows running these many tests only if wanted, so you can run the other benchmarks but not those. but I guess we could just use standard name based filtering instead?

File: packages/router-core/tests/route-interpolation.test.ts [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+test.each([
+  { path: '/static', params: {}, missing: false },
+  { path: '/$id', params: {}, missing: true },
+  { path: '/$id', params: { id: undefined }, missing: false },
+  { path: '/$id', params: { id: '' }, missing: false },
+  { path: '/{-$id}', params: {}, missing: false },
+  { path: '/pre{-$id}suffix', params: { id: null }, missing: false },
+  { path: '/files/$', params: {}, missing: true },
+  { path: '/files/$', params: { _splat: '' }, missing: true },
+  { path: '/files/{$}.txt', params: { _splat: 0 }, missing: true },
```
Comment (Sheraff) [inline packages/router-core/tests/route-interpolation.test.ts:23]: are we sure this should count as missing? `_splat: 0` is provided. It's probably an edge case, but it doesn't seem impossible that some people might use numbers as params

File: packages/router-core/tests/path-decoder.bench.ts (hunk 3 above)
Comment (Sheraff) [reply packages/router-core/tests/path-decoder.bench.ts:6]: yeah since we don't run benchmarks in CI, we never really run "all benchmarks" right? (just talking about `vitest bench` here, not codspeed obviously)

i personally would rather run them w/ name filtering, than have to remember that `TSR_LINK_PERF` is a thing and that sometimes i have to provide it
